File: trustly-android-sdk/src/main/java/net/trustly/android/sdk/util/CidManager.java
package net.trustly.android.sdk.util;

import android.annotation.SuppressLint;
import android.content.Context;
import android.provider.Settings;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class CidManager {

    private static final int EXPIRATION_TIME_LIMIT = 1;

    public static final String CID_PARAM = "CID";
    public static final String SESSION_CID_PARAM = "SESSION_CID";

    private CidManager() {
        throw new IllegalStateException("Utility class cannot be instantiated");
    }

    public static void generateCid(Context context) {
        CidStorage.saveData(context, CidStorage.CID, generateNewSession(context));
    }

    public static Map<String, String> getOrCreateSessionCid(Context context) {
        String cid = CidStorage.readDataFrom(context, CidStorage.CID);
        String sessionCid = CidStorage.readDataFrom(context, CidStorage.SESSION_CID);
        if (sessionCid == null) {
            sessionCid = cid;
        } else if (!isValid(sessionCid.split("-")[2])) {
            sessionCid = generateNewSession(context);
        }

        CidStorage.saveData(context, CidStorage.SESSION_CID, sessionCid);

        Map<String, String> values = new HashMap<>();
        values.put(CID_PARAM, cid);
        values.put(SESSION_CID_PARAM, sessionCid);
        return values;
    }

    private static String generateNewSession(Context context) {
        return getFingerPrint(context) + "-" + getRandomKey() + "-" + getTimestampBase36();
    }

    private static UUID getUUID() {
        return UUID.randomUUID();
    }

    @SuppressLint("HardwareIds")
    private static String getFingerPrint(Context context) {
        return Settings.Secure.getString(context.getContentResolver(),
                                  Settings.Secure.ANDROID_ID).substring(0, 4).toUpperCase();
    }

    private static String getRandomKey() {
        return getUUID().toString()
                .split("-")[2].toUpperCase();
    }

    private static String getTimestampBase36() {
        return Long.toString(Calendar.getInstance()
                                     .getTimeInMillis(), 36)
                .toUpperCase();
    }

    private static boolean isValid(String timestamp) {
        Calendar lastTime = Calendar.getInstance();
        lastTime.setTimeInMillis(Long.parseLong(timestamp, 36));
        return hoursAgo(lastTime) < EXPIRATION_TIME_LIMIT;
    }

    public static int hoursAgo(Calendar datetime) {
        Calendar now = Calendar.getInstance();
        long differenceInMillis = now.getTimeInMillis() - datetime.getTimeInMillis();
        long differenceInHours = (differenceInMillis) / (1000L * 60L * 60L);
        return (int) differenceInHours;
    }

}

File: trustly-android-sdk/src/main/java/net/trustly/android/sdk/util/cid/CidManager.kt
package net.trustly.android.sdk.util.cid

import android.annotation.SuppressLint
import android.content.Context
import android.provider.Settings
import net.trustly.android.sdk.util.cid.CidStorage.readDataFrom
import net.trustly.android.sdk.util.cid.CidStorage.saveData
import java.util.Calendar
import java.util.Locale
import java.util.UUID

object CidManager {

    private const val EXPIRATION_TIME_LIMIT: Int = 1

    const val CID_PARAM: String = "CID"
    const val SESSION_CID_PARAM: String = "SESSION_CID"
    private const val DIVIDER: String = "-"

    fun generateCid(context: Context) {
        saveData(context, CidStorage.CID, generateNewSession(context))
    }

    fun getOrCreateSessionCid(context: Context): Map<String, String?> {
        val cid = readDataFrom(context, CidStorage.CID)
        var sessionCid = readDataFrom(context, CidStorage.SESSION_CID)
        if (sessionCid == null) sessionCid = cid
        else {
            val split = sessionCid.split(DIVIDER)
            if (split.size > 2 && !isValid(split[2]))
                sessionCid = generateNewSession(context)
        }

        saveData(context, CidStorage.SESSION_CID, sessionCid)

        return hashMapOf(CID_PARAM to cid, SESSION_CID_PARAM to sessionCid)
    }

    private fun generateNewSession(context: Context) =
        getFingerPrint(context) + DIVIDER + getRandomKey() + DIVIDER + getTimestampBase36()

    private fun getUUID() = UUID.randomUUID()

    @SuppressLint("HardwareIds")
    private fun getFingerPrint(context: Context): String {
        return Settings.Secure.getString(
            context.contentResolver,
            Settings.Secure.ANDROID_ID
        ).substring(0, 4).uppercase(Locale.getDefault())
    }

    private fun getRandomKey() =
        getUUID().toString().split(DIVIDER.toRegex())[2].uppercase(Locale.getDefault())

    private fun getTimestampBase36() =
        Calendar.getInstance().timeInMillis.toString(36).uppercase(Locale.getDefault())

    private fun isValid(timestamp: String): Boolean {
        val lastTime = Calendar.getInstance()
        lastTime.timeInMillis = timestamp.toLong(36)
        return hoursAgo(lastTime) < EXPIRATION_TIME_LIMIT
    }

    private fun hoursAgo(datetime: Calendar): Int {
        val now = Calendar.getInstance()
        val differenceInMillis = now.timeInMillis - datetime.timeInMillis
        val differenceInHours = (differenceInMillis) / (1000L * 60L * 60L)
        return differenceInHours.toInt()
    }

}